2019/intcode.d

```diff
@@ -1,66 +1,90 @@
 module intcode;
 
-struct IntCode {
-  int[] memory;
-  int[] input;
-  int[] output;
+alias IntCode = IntCodeT!int;
+alias LongCode = IntCodeT!long;
 
-  alias Addr = int;
+struct IntCodeT(T) {
+  alias Word = T;
 
-  Addr PC;
+  Word[] memory;
+  Word[] input;
+  Word[] output;
+
+  alias Addr = Word;
+
+  Addr PC, RB;
   bool halted = true;
   bool feedback = false;
 
-  enum Opcode { Add = 1, Mul = 2, In = 3, Out = 4, JT = 5, JF = 6, Lt = 7, Eq = 8, End = 99 }
+  enum Opcode {
+    Add = 1, Mul
+  , In = 3, Out
+  , JT = 5, JF
+  , Lt = 7, Eq
+  , Rel = 9
+  , End = 99
+  }
 
-  struct Instr { Opcode op; Addr ld1, ld2, st; }
+  enum Mode { Pos, Imm, Rel }
 
   bool execute() {
     import std.conv, std.range;
 
     auto opcode = memory[PC];
-    bool p1imm = (opcode / 100) % 10 == 1;
-    bool p2imm = (opcode / 1000) % 10 == 1;
-
-    int v(int n, bool imm) {
-      return imm ? memory[n] : memory[memory[n]];
+    Mode m1 = to!Mode((opcode / 100) % 10);
+    Mode m2 = to!Mode((opcode / 1000) % 10);
+    Mode m3 = to!Mode((opcode / 10000) % 10);
+
+    Word ld(Addr n, Mode mode) {
+      auto param = memory[PC+n];
+
+      final switch (mode) {
+        case Mode.Pos: if (memory.length <= param) memory.length = param+1; return memory[param];
+        case Mode.Imm: return param;
+        case Mode.Rel: if (memory.length <= RB+param) memory.length = RB+param+1; return memory[RB+param];
+      }
     }
 
-    Instr instr;
+    void st(Addr n, Mode mode, Word v) {
+      auto param = memory[PC+n];
 
-    void ld() {
-      instr.ld1 = v(PC+1, p1imm);
-      instr.ld2 = v(PC+2, p2imm);
+      final switch (mode) {
+        case Mode.Pos: if (memory.length <= param) memory.length = param+1; memory[param] = v; return;
+        case Mode.Imm: return;
+        case Mode.Rel: if (memory.length <= RB+param) memory.length = RB+param+1; memory[RB+param] = v; return;
+      }
     }
 
-    instr.op = to!Opcode(opcode % 100);
+    Opcode op = to!Opcode(opcode % 100);
 
-    final switch (instr.op) {
-      case Opcode.JT:  ld(); PC = (instr.ld1 != 0) ? instr.ld2 : (PC + 3); return true;
-      case Opcode.JF:  ld(); PC = (instr.ld1 == 0) ? instr.ld2 : (PC + 3); return true;
-      case Opcode.Add: ld(); memory[memory[PC+3]] = instr.ld1 + instr.ld2; PC = PC + 4; return true;
-      case Opcode.Mul: ld(); memory[memory[PC+3]] = instr.ld1 * instr.ld2; PC = PC + 4; return true;
-      case Opcode.Lt:  ld(); memory[memory[PC+3]] = instr.ld1 < instr.ld2; PC = PC + 4; return true;
-      case Opcode.Eq:  ld(); memory[memory[PC+3]] = instr.ld1 == instr.ld2; PC = PC + 4; return true;
+    final switch (op) {
+      case Opcode.JT:  PC = (ld(1, m1) != 0) ? ld(2, m2) : (PC + 3); return true;
```

Comment:
Although this looks clean, the [compiler](https://gcc.godbolt.org/z/RwGPfh) is generating actual function calls here which can't be helping performance. The modal memory access code is duplicated. Is there a way to ask the runtime for a memory allocator that takes care of this for you?

Reply:
`std.container.array` looks promising but would expose an interface change (`IntCode.memory` is no longer an array).